common/devdoc/gballoc_ll_mimalloc.md

```diff
@@ -0,0 +1,90 @@
+# gballoc_ll_mimalloc requirements
+================
+
+## Overview
+
+gballoc_ll_mimalloc is a module that delegates all call of its APIs to the ones from mimalloc.
+
+## References
+[mimalloc](https://github.com/microsoft/mimalloc)
+
+
+## Exposed API
+
+```c
+    MOCKABLE_FUNCTION(, int, gballoc_ll_init, void*, params);
+    MOCKABLE_FUNCTION(, void, gballoc_ll_deinit);
+
+    MOCKABLE_FUNCTION(, void*, gballoc_ll_malloc, size_t, size);
+    MOCKABLE_FUNCTION(, void, gballoc_ll_free, void*, ptr);
+    MOCKABLE_FUNCTION(, void*, gballoc_ll_calloc, size_t, nmemb, size_t, size);
+    MOCKABLE_FUNCTION(, void*, gballoc_ll_realloc, void*, ptr, size_t, size);
+```
+
+### gballoc_ll_init
+```c
+MOCKABLE_FUNCTION(, int, gballoc_ll_init, void*, params);
+```
+
+`gballoc_ll_init` returns. `params` exists as a placeholder and is ignored.
+
+**SRS_GBALLOC_LL_MIMALLOC_02_001: [** `gballoc_ll_init` shall return 0. **]**
+
+
+### gballoc_ll_deinit
+```c
+MOCKABLE_FUNCTION(, void, gballoc_ll_deinit);
+```
+
+`gballoc_ll_deinit` returns.
+
+**SRS_GBALLOC_LL_MIMALLOC_02_002: [** `gballoc_ll_deinit` shall return. **]**
+
+### gballoc_ll_malloc
+```c
+MOCKABLE_FUNCTION(, void*, gballoc_ll_malloc, size_t, size);

```

Comment:
This is a golden opportunity to log the size of buffer to be allocated when this fails .. rather than logging it in every module that uses malloc. I'll buy you a coffee and an extra bike tire if you fix this :smile:

Reply:
I believe you're selling yourself cheap here because printing the size here bring little information. All it would print would be "failed to allocate 20 bytes". Now... what's the meaning of these 20 bytes, and where do they come from, that is left as an exercise to the user... I mean without a call stack / previous/next messages the message is meaningless. And even when you'd have the previous/next LogErrors, the context where it was allocated still matter. I suppose it is better to keep all the information in one LogError, but... 

I need that tire. I think I would like to try Conti Gatorskins :)

So. Added logerrors here and everywhere... when is my tire coming? :)

---
In reply to: [457679818](https://github.com/Azure/azure-c-pal/pull/11#discussion_r457679818) [](ancestors = 457679818)

Reply:
The user will log all failures in their code anyway, so it should be a pretty easy thread to follow. 
For e.g. malloc logs -> malloc 12345678910 bytes failed
User logs  "malloc failed" .. put two together and you have your data.